kv/local_sender.go

```diff
@@ -142,13 +142,28 @@ func (ls *LocalSender) Send(call client.Call) {
 // by consulting each store in turn via Store.LookupRange(key).
 // Returns RaftID and replica on success; RangeKeyMismatch error
 // if not found.
-func (ls *LocalSender) lookupReplica(start, end proto.Key) (int64, *proto.Replica, error) {
+// TODO(tschottdorf) with a very large number of stores, the LocalSender
+// may want to avoid scanning the whole map of stores on each invocation.
+func (ls *LocalSender) lookupReplica(start, end proto.Key) (raftID int64, replica *proto.Replica, err error) {
 	ls.mu.RLock()
 	defer ls.mu.RUnlock()
+	var rng *storage.Range
 	for _, store := range ls.storeMap {
-		if rng := store.LookupRange(start, end); rng != nil {
-			return rng.Desc().RaftID, rng.GetReplica(), nil
+		rng = store.LookupRange(start, end)
+		if rng == nil {
+			continue
 		}
+		if replica == nil {
+			raftID = rng.Desc().RaftID
+			replica = rng.GetReplica()
+			continue
+		}
+		// Should never happen outside of tests.
+		return 0, nil, util.Errorf(
```

Comment:
this is a sanity check as there were tests that violated this. The replication queue is making sure to never put more than one copy of each replica on a node, and the store.AddReplica checks it again (see this PR).